### PR TITLE
New post: Symbolic operators and type classes for Cats

### PR DESCRIPTION
**Pull request by larsrh**
Original author: @non
Source: https://gist.github.com/non/3abdb35a72c39276d3d9

### CODE REVIEW

**Comment by larsrh** [comment]
The second and final one. @non, I hope I didn't destroy the narrative structure of your posts.

**Comment by non** [comment]
loooks good to me :+1:

File: _posts/2015-08-07-symbolic-operators.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+This post is an introduction into how operators are implemented in Cats and has been originally published in [August 2015](https://gist.github.com/non/3abdb35a72c39276d3d9).
+Some more details can be found in the [previous post]({% post_url 2015-08-06-machinist %}).
+
+One of the simplest and most recognizable type classes is the semigroup.
+This type class abstracts over the ability to combine values of a certain
+type in an associative manner.
+
+<div class="side-note">
+  What does <em>associativity</em> mean?
+  We call an operation `⊕` associative, if for all `a`, `b` and `c`, `a ⊕ (b ⊕ c) = (a ⊕ b) ⊕ c` holds.
```
Comment (S11001001) [inline _posts/2015-08-07-symbolic-operators.md]: [Even backticks do not work in side-note.](https://github.com/typelevel/typelevel.github.com/pull/47#issuecomment-124268180)

Also, can you see those symbols @tpolecat?

File: _posts/2015-08-07-symbolic-operators.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+method, which returns a type that has a `|+|` method (specifically
+[`SemigroupOps[A]`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/core/src/main/scala/cats/syntax/semigroup.scala#L12)).
+However, `semigroupSyntax` requires an implicit `Semigroup[A]` value to be in scope.
+Do we have a `Semigroup[Int]` in scope?
+
+Yes we do. Our import also provides an implicit value [`intGroup`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/std/src/main/scala/cats/std/anyval.scala#L23)
+of type `AdditiveCommutativeGroup[Int]`. Leaving aside what *additive*, *commutative*,
+and *group* mean here, this is a subtype of `Semigroup[Int]`, so it matches.
+
+Let's continue with our current example. At this point we have gone form:
```
Comment (S11001001) [inline _posts/2015-08-07-symbolic-operators.md]: s/form:/from:/

File: _posts/2015-08-07-symbolic-operators.md (hunk 1 above)
Comment (tpolecat) [inline _posts/2015-08-07-symbolic-operators.md]: Yep, crosshairs.

File: _posts/2015-08-07-symbolic-operators.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+semigroupSyntax[Int](19)(intGroup) |+| 20
+```
+
+into:
+
+```scala
+intGroup.combine(19, 20)
+```
+
+The suggestive map item tells us that we should rewrite the `|+|` operator
```
Comment (S11001001) [inline _posts/2015-08-07-symbolic-operators.md]: Maybe, "The *aforementioned* suggestive map item", or some other connective tissue to tie it to the `operatorNames` reference above?

File: _posts/2015-08-07-symbolic-operators.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+You may also decide that the syntax convenience is not worth it. To write our
+original example without syntax implicits (but still using implicit values)
+you could say:
+
+```scala
+import cats.Semigroup
+import cats.implicits._
+
+def gsum[A](values: List[A])(implicit s: Semigroup[A]): Option[A] =
+  if (values.isEmpty) None else values.reduceLeft((x, y) => s.combine(x, y0))
```
Comment (non) [inline _posts/2015-08-07-symbolic-operators.md]: I just found this bug, `y0` should be `y`.

File: _posts/2015-08-07-symbolic-operators.md (hunk 3 above)
Comment (non) [inline _posts/2015-08-07-symbolic-operators.md]: +1

File: _posts/2015-08-07-symbolic-operators.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+as we would expect.
+
+Whew!
+
+Conclusion
+----------
+
+This is a lot of machinery. The incredibly terse and expressive syntax it
+enables is quite nice, but you can see that even leaving out one import
+will cause the whole edifice to come tumbling down.
```
Comment (S11001001) [inline _posts/2015-08-07-symbolic-operators.md]: I would say a little more here: ...come tumbling down, as so often occurs when such things are constructed from Scala implicits.  In this case, we needed an implicit conversion, *and* an entirely separate implicit value, just to make a simple "add" operation work.

But that is more subjective than the other things I've commented on.

Comment (non) [inline _posts/2015-08-07-symbolic-operators.md]: Heh, well, I think there are two perspectives.

From a language with first class type classes, yes, I agree with your assessment.

From any other language (especially those without any way to create generic operators), it's a lot better than the status quo despite these problems.

Comment (S11001001) [inline _posts/2015-08-07-symbolic-operators.md]: Yes, I would say that unfortunately in Scala this is the best choice.  We should still say what the shortcomings of the best choice are, lest someone think everything is hunky dory.

**Comment by larsrh** [comment]
@S11001001 @non 2c55fb1

File: _posts/2015-08-07-symbolic-operators.md (hunk 5 above)
Comment (larsrh) [inline _posts/2015-08-07-symbolic-operators.md]: I've left out this suggestion, @S11001001. Since it's @non's text, I'll let him decide.

**Pull request merged**